src/librustdoc/json/conversions.rs

```diff
@@ -105,37 +102,37 @@ impl JsonRenderer<'_> {
             Some(ty::Visibility::Public) => Visibility::Public,
             Some(ty::Visibility::Restricted(did)) if did.is_crate_root() => Visibility::Crate,
             Some(ty::Visibility::Restricted(did)) => Visibility::Restricted {
-                parent: id_from_item_default(did.into(), self.tcx),
-                path: self.tcx.def_path(did).to_string_no_crate_verbose(),
+                parent: self.cache.intern_item_id(ItemId::type_ns(did)),
+                path: self.cache.tcx.def_path(did).to_string_no_crate_verbose(),
             },
         }
     }
 }
 
-pub(crate) trait FromWithTcx<T> {
-    fn from_tcx(f: T, tcx: TyCtxt<'_>) -> Self;
+pub(crate) trait FromWithCache<T> {
+    fn from_with_cache(f: T, cache: &Cache<'_>) -> Self;
```

Comment:
I don't think this is the right approach to threading some mutable state through the clean->JSON pass. For one, it hides the fact that `Cache` is mutable, why doesn't it take `&mut Cache`, so we can remove the `Lock` on the ItemId interning map.

But more broadly, I don't think we should be adding more things to `Cache` that are only needed here. I think a better aproach would be to do

```rust
// TODO: Better name for this
struct JsonItemConverter<'tcx> {
   tcx: TyCtxt<'tcx>,
   cache: Cache,
   id_interner: FxHashMap<ItemId, rustdoc_json_types::Id>
}
```

and then pass that to all the converting methods (either via a trait, or via methods on `impl JsonItemConverter`